File: 2023/js/day14.js
import { forEachInGrid, formatTime, printGrid, runDay } from '../../utils.js';

// https://adventofcode.com/2023/day/14

runDay(2023, 14)
  .part(1, part1)
  .part(2, part2)
  .end();

function part1(inp) {
  let grid = inp.split('\n').map(l => l.split(''));
  let round = 'O';

  forEachInGrid(grid, (cell, i, j) => {
    if (cell === round) {
      grid[i][j] = '.';
      while (grid[i - 1]?.[j] === '.') {
        i--;
      }
      grid[i][j] = round;
    }
  });
  let height = grid.length;

  let total = 0;
  forEachInGrid(grid, (cell, i) => {
    if (cell === round) {
      total += height - i;
    }
  });

  return total;
}

function part2(inp) {
  let grid = inp.split('\n').map(l => l.split(''));
  let rock = 'O';
  let limit = 1_000_000_000;

  let gridsHistory = {};
  let loopFound = false;

  let width = grid[0].length;
  let height = grid.length;

  for (let i = 1; i <= limit; i++) {
    // to top
    for (let j = 0; j < width; j++) {
      let jSlow = 0;
      let jFast = 0;

      while (jFast < height) {
        while (jSlow < height - 1 && grid[jSlow][j] !== '.') {
          jSlow++;
        }
        let l = jSlow;

        jFast = Math.max(jSlow + 1, jFast);

        while (grid[jFast]?.[j]) {
          if (grid[jFast][j] === rock) break;
          if (grid[jFast][j] === '#') {
            jFast++;
            jSlow = jFast;
            l = jFast;
            break;
          }

          jFast++;
        }
        let r = jFast;

        if (grid[l]?.[j] === '.' && grid[r]?.[j] === rock) {
          grid[l][j] = rock;
          grid[r][j] = '.';
          jSlow++;
          jFast++;
        }
      }
    }

    // to left
    for (let i = 0; i < height; i++) {
      let row = grid[i];
      let jSlow = 0;
      let jFast = 0;

      while (jFast < width) {
        while (jSlow < height && row[jSlow] !== '.') {
          jSlow++;
        }
        let slow = jSlow;

        jFast = Math.max(jSlow + 1, jFast);

        while (row[jFast]) {
          if (row[jFast] === rock) break;
          if (row[jFast] === '#') {
            jFast++;
            jSlow = jFast;
            slow = jFast;
            break;
          }

          jFast++;
        }
        let fast = jFast;

        if (row[slow] === '.' && row[fast] === rock) {
          row[slow] = rock;
          row[fast] = '.';
          jSlow++;
          jFast++;
        }
      }
    }

    // to bottom
    for (let j = 0; j < width; j++) {
      let jFast = height - 1;
      let jSlow = height - 1;

      while (jFast >= 0) {
        while (jSlow > 0 && grid[jSlow][j] !== '.') {
          jSlow--;
        }
        let slow = jSlow;

        jFast = Math.min(jSlow - 1, jFast);

        while (jFast >= 0) {
          if (grid[jFast][j] === rock) break;
          if (grid[jFast][j] === '#') {
            jSlow = jFast;
            slow = jFast;
            break;
          }

          jFast--;
        }
        let fast = jFast;

        if (grid[fast]?.[j] === rock && grid[slow]?.[j] === '.') {
          grid[fast][j] = '.';
          grid[slow][j] = rock;
          jFast--;
          jSlow--;
        }
      }
    }

    // to right
    for (let i = 0; i < height; i++) {
      let row = grid[i];
      let jFast = width - 1;
      let jSlow = width - 1;

      while (jFast >= 0) {
        while (jSlow > 0 && row[jSlow] !== '.') {
          jSlow--;
        }
        let slow = jSlow;

        jFast = Math.min(jSlow - 1, jFast);

        while (jFast >= 0) {
          if (row[jFast] === rock) break;
          if (row[jFast] === '#') {
            jSlow = jFast;
            slow = jFast;
            break;
          }

          jFast--;
        }
        let fast = jFast;

        if (row[fast] === rock && row[slow] === '.') {
          row[fast] = '.';
          row[slow] = rock;
          jFast--;
          jSlow--;
        }
      }
    }

    if (!loopFound) {
      let currentGrid = printGrid(grid);
      if (gridsHistory[currentGrid]) {
        loopFound = true;

        let loopLen = i - gridsHistory[currentGrid];
        let cyclesCnt = Math.floor((limit - gridsHistory[currentGrid]) / loopLen);

        i = gridsHistory[currentGrid] + loopLen * cyclesCnt;
      } else {
        gridsHistory[currentGrid] = i;
      }
    }
  }

  let total = 0;
  forEachInGrid(grid, (cell, i) => {
    if (cell === rock) {
      total += height - i;
    }
  });

  return total;
}

File: rust/src/y2023/day14.rs
pub fn part1(input: &str) -> i32 {
    5
}

/*

JS
progress 0.020% [ '0:57.065' ] 0.2kk
progress 0.021% [ '0:59.924' ] 0.21kk
progress 0.022% [ '1:02.781' ] 0.22kk
progress 0.023% [ '1:05.643' ] 0.23kk
// 4347 min = 72h

RUST
progress: 0.144% time: 58:278 1.44kk
progress: 0.146% time: 59:  8 1.46kk
progress: 0.148% time: 59:754 1.48kk
progress: 0.15% time: 60:516 1.5kk
progress: 0.152% time: 61:326 1.52kk
progress: 0.154% time: 62: 74 1.54kk
// 666 min = 11h

*/

pub fn part2(input: &str) -> usize {
    let mut grid: Vec<Vec<u8>> = input.lines().map(|l| l.bytes().collect()).collect();

    let rock = b'O';
    let limit = 1_000_000_000;
    // let limit = 1_000;

    let width = grid[0].len();
    let height = grid.len();

    let start = std::time::Instant::now();

    for i in 1..=limit {
        // if (i % 20_000 == 0) {
        //     println!(
        //         "progress: {:3}% time: {}:{:3} {}kk",
        //         ((i * 100) as f32 / (limit as f32)),
        //         start.elapsed().as_secs(),
        //         start.elapsed().as_millis() % 1000,
        //         (i as f32) / 1_000_000.0
        //     )
        // }

        // to top
        for j in 0..width {
            let mut j_slow = 0;
            let mut j_fast = 0;

            while j_fast < height {
                while j_slow < height - 1 && grid[j_slow][j] != b'.' {
                    j_slow += 1;
                }
                let mut slow = j_slow;

                j_fast = usize::max(j_slow + 1, j_fast);

                while j_fast < height {
                    if grid[j_fast][j] == rock {
                        break;
                    }
                    if grid[j_fast][j] == b'#' {
                        j_fast += 1;
                        j_slow = j_fast;
                        slow = j_fast;
                        break;
                    }

                    j_fast += 1;
                }

                let fast = j_fast;

                if slow < height && fast < height && grid[slow][j] == b'.' && grid[fast][j] == rock
                {
                    grid[slow][j] = rock;
                    grid[fast][j] = b'.';
                    j_slow += 1;
                    j_fast += 1;
                }
            }
        }
        // println!("top");

        // to left
        for j in 0..height {
            let mut j_slow = 0;
            let mut j_fast = 0;

            while j_fast < width {
                while j_slow < height && grid[j][j_slow] != b'.' {
                    j_slow += 1;
                }
                let mut slow = j_slow;

                j_fast = usize::max(j_slow + 1, j_fast);

                while j_fast < height {
                    if grid[j][j_fast] == rock {
                        break;
                    }
                    if grid[j][j_fast] == b'#' {
                        j_fast += 1;
                        j_slow = j_fast;
                        slow = j_fast;
                        break;
                    }

                    j_fast += 1;
                }

                let fast = j_fast;

                if slow < width && fast < width && grid[j][slow] == b'.' && grid[j][fast] == rock {
                    grid[j][slow] = rock;
                    grid[j][fast] = b'.';
                    j_slow += 1;
                    j_fast += 1;
                }
            }
        }
        // println!("left");

        // to bottom
        for j in 0..width {
            let mut j_slow = 0;
            let mut j_fast = 0;

            while j_fast < height {
                while j_slow < height && grid[j_slow][j] != b'.' {
                    j_slow += 1;
                }
                let mut slow = j_slow;

                j_fast = usize::max(j_slow + 1, j_fast);

                while j_fast < height {
                    if grid[j_fast][j] == rock {
                        break;
                    }
                    if grid[j_fast][j] == b'#' {
                        j_fast += 1;
                        j_slow = j_fast;
                        slow = j_fast;
                        break;
                    }

                    j_fast += 1;
                }

                let fast = j_fast;

                if slow < height && fast < height && grid[slow][j] == b'.' && grid[fast][j] == rock
                {
                    grid[slow][j] = rock;
                    grid[fast][j] = b'.';
                    j_slow += 1;
                    j_fast += 1;
                }
            }
        }
        // println!("bottom");

        // to right
        for j in 0..height {
            let mut j_slow = 0;
            let mut j_fast = 0;

            while j_fast < width {
                while j_slow < height && grid[j][j_slow] != b'.' {
                    j_slow += 1;
                }
                let mut slow = j_slow;

                j_fast = usize::max(j_slow + 1, j_fast);

                while j_fast < width {
                    if grid[j][j_fast] == rock {
                        break;
                    }
                    if grid[j][j_fast] == b'#' {
                        j_fast += 1;
                        j_slow = j_fast;
                        slow = j_fast;
                        break;
                    }

                    j_fast += 1;
                }

                let fast = j_fast;

                if slow < width && fast < width && grid[j][slow] == b'.' && grid[j][fast] == rock {
                    grid[j][slow] = rock;
                    grid[j][fast] = b'.';
                    j_slow += 1;
                    j_fast += 1;
                }
            }
        }
        // println!("right");
    }

    let mut total = 0;

    for i in 0..height {
        for j in 0..width {
            if grid[i][j] == rock {
                total += height - i;
            }
        }
    }

    total
}
